eclair-core/src/main/scala/fr/acinq/eclair/io/PeerReadyNotifier.scala

```diff
@@ -0,0 +1,196 @@
+    }
+    // We check whether the peer is already connected.
+    switchboard ! Switchboard.GetPeerInfo(peerInfoAdapter, remoteNodeId)
+    Behaviors.receiveMessagePartial {
+      case PeerNotConnected =>
+        context.log.debug("peer is not connected yet")
+        Behaviors.same
+      case SomePeerConnected(nodeId) =>
+        if (nodeId == remoteNodeId) {
+          switchboard ! Switchboard.GetPeerInfo(peerInfoAdapter, remoteNodeId)
+        }
+        Behaviors.same
+      case PeerChannels(channels) =>
+        if (channels.isEmpty) {
+          context.log.info("peer is ready with {} channels", channels.size)
+          replyTo ! PeerReady(remoteNodeId, 0)
+          Behaviors.stopped
+        } else {
+          context.log.debug("peer is connected with {} channels", channels.size)
+          waitForChannelsReady(replyTo, remoteNodeId, channels, context, timers)
+        }
+      case NewBlockNotTimedOut(currentBlockHeight) =>
+        context.log.debug("waiting for peer to connect at block {}", currentBlockHeight)
+        Behaviors.same
+      case Timeout =>
+        context.log.info("timed out waiting for peer to be ready")
+        replyTo ! PeerUnavailable(remoteNodeId)
+        Behaviors.stopped
+    }
+  }
+
+  def waitForChannelsReady(replyTo: ActorRef[Result], remoteNodeId: PublicKey, channels: Set[akka.actor.ActorRef], context: ActorContext[Command], timers: TimerScheduler[Command]): Behavior[Command] = {
+    timers.startTimerWithFixedDelay(CheckChannelsReady, initialDelay = 50 millis, delay = 1 second)
+    Behaviors.receiveMessagePartial {
+      case CheckChannelsReady =>
+        context.spawnAnonymous(ChannelStatesCollector(context.self, channels))
+        Behaviors.same
+      case ChannelStates(states) =>
+        if (states.forall(isChannelReady)) {
+          replyTo ! PeerReady(remoteNodeId, channels.size)
+          Behaviors.stopped
+        } else {
+          context.log.debug("peer has {} channels that are not ready", states.count(s => !isChannelReady(s)))
+          Behaviors.same
+        }
+      case NewBlockNotTimedOut(currentBlockHeight) =>
+        context.log.debug("waiting for channels to be ready at block {}", currentBlockHeight)
+        Behaviors.same
+      case SomePeerConnected(_) =>
+        Behaviors.same
+      case Timeout =>
+        context.log.info("timed out waiting for channels to be ready")
+        replyTo ! PeerUnavailable(remoteNodeId)
+        Behaviors.stopped
+    }
+  }
+
+  // We use an exhaustive pattern matching here to ensure we explicitly handle future new channel states.
+  private def isChannelReady(state: channel.ChannelState): Boolean = state match {
+    case channel.WAIT_FOR_INIT_INTERNAL => false
+    case channel.WAIT_FOR_INIT_SINGLE_FUNDED_CHANNEL => false
+    case channel.WAIT_FOR_INIT_DUAL_FUNDED_CHANNEL => false
+    case channel.OFFLINE => false
+    case channel.SYNCING => false
+    case channel.WAIT_FOR_OPEN_CHANNEL => true
```

Comment:
For async-payments, what happens if `isChannelReady`  is true, but the channel is in a state before  `NORMAL` ? Would the channel logic queue the payment until the channel reaches `NORMAL`?  or would it fail the payment?

For `SHUTDOWN` or later I assume it would be correct to trigger and then fail an async-payment.

Reply:
We only care about verifying that the channels are fully initialized here, that's what guarantees that the peer is "ready". The actual state of those channels is unrelated to whether the peer is ready or not, and should be handled by other components (for payments, path-finding will deal with whether we can relay the payment or not, I don't think we should hold the payment waiting for a channel to confirm, at least not in those actors).

Reply:
Your comment now explains it well; it only waits until channels have finished figuring out what state they're actually in, but no need to wait for any particular state.